src/trusted-device/Security/TwoFactor/Trusted/JwtTokenEncoder.php

```diff
@@ -23,18 +23,17 @@ class JwtTokenEncoder
     public const CLAIM_FIREWALL = 'fwl';
     public const CLAIM_VERSION = 'vsn';
 
-    private Configuration $configuration;
+    private Clock $clock;
 
-    public function __construct(string $applicationSecret)
+    public function __construct(private Configuration $configuration, ?Clock $clock = null)
     {
-        $this->configuration = Configuration::forSymmetricSigner(new Sha256(), InMemory::plainText($applicationSecret));
-        $this->configuration->setValidationConstraints(new SignedWith($this->configuration->signer(), $this->configuration->signingKey()));
+        $this->clock = $clock ?? SystemClock::fromSystemTimezone();
     }
 
     public function generateToken(string $username, string $firewallName, int $version, DateTimeImmutable $validUntil): Plain
     {
         $builder = $this->configuration->builder()
-            ->issuedAt(new DateTimeImmutable())
+            ->issuedAt($this->clock->now())
```

Comment:
What's the advantage of using `Clock` here, over using a DateTimeImmutable?

And wouldn't it be necessary to declare a dependency to `lcobucci/clock` in the composer.json, since we're directly referencing code from that component?

Reply:
https://blog.frankdejonge.nl/being-in-control-of-time-in-php/ has some info on the motivation for this. We're also working on the psr-20 to promote interoperability between different implementations.

I think that adding the dependency is up to your policy regarding direct/indirect dependencies. I'd probably add it if it was my library 👍

Reply:
Ok, I get the point of encapsulating the question "what time is now", which is really helpful for the matter of testing. ~Though in this specific context I don't really see the benefit of pulling in another component/dependency, especially since we're not using its capabilities in the tests. It could be a `DateTimeImmutable` and then I wouldn't need to deal with another dependency.~

~If a standardized clock interface is becoming a thing, I'd rather brush over the entire bundle once, to replace all time-dependencies with a clock implementation and have it aligned across the entire library.~

^ Keeping this to document my though process.

Just realized that `Constraint\ValidAt` is depending on `Lcobucci\Clock\Clock`, so my idea of "just use a `DateTimeImmutable`" is obviously not feasible. In that case: Since we're referencing components from `lcobucci/clock` in the 2fa-bundle code, imo it has to go into the `composer.json`, because I cannot necessarily rely on 3rd party packages to pull in the dependencies that I'm referring in my code.